### ISSUE TITLE
Run `.chezmoiexternal` after `run_before` but before everything else

### ISSUE THREAD

**Issue by felipecrs** [twpayne/chezmoi#1616]
## Is your feature request related to a problem? Please describe.

`chezmoiexternal`s provides a way to fetch files that are not in the source state. However, during templates, we might want to deal with some of these externally defined files, and we need to assume that these files would be created before the other templates get executed.

For the rest of the files, they would indeed exist before anything else, as they would be source state. This is the reason why this exception should be applied for `chezmoiexternal`.

## Describe the solution you'd like

In order for that to work, chezmoiexternal should run before everything else (but still after `run_before`).

## Describe alternatives you've considered

It's lucky that alphabetically chezmoiexternal is already one of the first letters, so for most cases this would not be noticed.

## Additional context

Here is an example:

```yaml
# .chezmoiexternal.yaml

".deno/bin/deno":
  type: file
  url: "https://github.com/denoland/deno/releases/download/v1.16.1/deno-x86_64-unknown-linux-gnu.zip"
  executable: true
  filter:
    command: zcat
```

Here is a template which would depend on it:

```
{{- output (joinPath .chezmoi.homeDir ".deno/bin/deno") "completions" "zsh" -}}
```

But as mentioned, luckly, this doesn't trigger the problem.

**Comment by felipecrs** [comment]
@twpayne I think this is a clean a way to tackle https://github.com/twpayne/chezmoi/issues/1574.

If externals are guaranteed to be executed before, whatever is applied after does not take into account during `chezmoiexternal`s resolving.

**Comment by twpayne** [comment]
This will require some careful thought.

Currently chezmoi has the model (1) read source state (including reading externals) (2) compute target state (3) apply target state.

Running scripts is part of step 3, and the reason that the example you give doesn't trigger the problem is because chezmoi computes the target state lazily (i.e. only when needed) in step 3, so the template happens to only be executed after the external is applied (i.e. after `deno` is installed).

What this issue is effectively proposing is a new model of (1) read source state excluding externals (2) compute target state excluding externals (2) run `run_before_` scripts (3) change source state by reading externals (4) compute target state including externals (and hope that nothing important changed) (5) apply new target state.

Basically, the proposal here breaks the core chezmoi assumption "the source state is never modified while chezmoi is running". Making chezmoi behave correctly when that core assumption is broken is likely to be extremely hard.

**Comment by felipecrs** [comment]
> the proposal here breaks the core chezmoi assumption "the source state is never modified while chezmoi is running"

I don't think this applies to here. The `deno` binary was never part of my real dotfiles repository. It only gets fetched by `.chezmoiexternal`. So, in other words, in order to compute the real status of the source state, in which some files would be missing because they are "external", I think it still makes sense to add precedence to `.chezmoiexternal`.
